File: apidiff/render_json.py
"""Machine-readable diff artifact.

Key order is fixed and every collection is sorted so the Python and Node ports produce byte-identical
files — that equality is the repo's dual-port parity gate.
"""

import json

from .model import SEVERITY_ORDER


def build(result) -> dict:
    """Return the JSON-serializable form of a diff."""
    types = {}
    for delta in result.deltas:
        bucket = types.setdefault(delta.type_uid, {"added": [], "removed": [], "changed": []})
        bucket[delta.kind].append({
            "uid": delta.uid,
            "display": delta.display,
            "severity": delta.severity,
            "target": delta.target,
            "before": delta.before,
            "after": delta.after,
            "reasons": list(delta.reasons),
        })

    return {
        "product": result.product_code,
        "productName": result.product_name,
        "from": result.version_from,
        "to": result.version_to,
        "generatedFrom": "xrefmap+html",
        "summary": {
            **result.summary(),
            "total": len(result.deltas),
            "typesFrom": len(result.surface_from.types) if result.surface_from else 0,
            "typesTo": len(result.surface_to.types) if result.surface_to else 0,
        },
        "severities": list(SEVERITY_ORDER),
        "types": [
            {"fqn": fqn, **{kind: types[fqn][kind] for kind in ("added", "removed", "changed")}}
            for fqn in sorted(types)
        ],
        "warnings": sorted(result.warnings),
    }


def render(result) -> str:
    """Serialize a diff as pretty-printed JSON with a trailing newline."""
    return json.dumps(build(result), indent=2, ensure_ascii=False) + "\n"

File: apidiff/render-json.mjs
/**
 * render-json.mjs — machine-readable diff artifact (Node port of render_json.py).
 *
 * Key order is fixed and every collection is sorted so this and the Python port produce
 * byte-identical files — that equality is the repo's dual-port parity gate.
 */

import { SEVERITY_ORDER, summary } from "./model.mjs";

/** Return the JSON-serializable form of a diff. */
export function build(result) {
  const types = new Map();
  for (const delta of result.deltas) {
    if (!types.has(delta.typeUid)) types.set(delta.typeUid, { added: [], removed: [], changed: [] });
    types.get(delta.typeUid)[delta.kind].push({
      uid: delta.uid,
      display: delta.display,
      severity: delta.severity,
      target: delta.target,
      before: delta.before,
      after: delta.after,
      reasons: [...delta.reasons],
    });
  }

  const compareStrings = (a, b) => (a < b ? -1 : a > b ? 1 : 0);

  return {
    product: result.productCode,
    productName: result.productName,
    from: result.versionFrom,
    to: result.versionTo,
    generatedFrom: "xrefmap+html",
    summary: {
      ...summary(result),
      total: result.deltas.length,
      typesFrom: result.surfaceFrom ? result.surfaceFrom.types.size : 0,
      typesTo: result.surfaceTo ? result.surfaceTo.types.size : 0,
    },
    severities: [...SEVERITY_ORDER],
    types: [...types.keys()].sort(compareStrings).map((fqn) => ({
      fqn,
      added: types.get(fqn).added,
      removed: types.get(fqn).removed,
      changed: types.get(fqn).changed,
    })),
    warnings: [...result.warnings].sort(compareStrings),
  };
}

/** Serialize a diff as pretty-printed JSON with a trailing newline. */
export function render(result) {
  return `${JSON.stringify(build(result), null, 2)}\n`;
}
